File: projects/Vibe_XASR/macos_build/native/app/Sources/VibeIME/Refiner.swift
import Foundation

// AI 润色(Beta)—— 后端无关的骨架。
// =========================================================================
// 设计:
//   · 后端(llama.cpp / MLX / 子进程)实现 `RefinerBackend`,只管"文本进 → 润色出"。
//   · `Refiner` 门面协调"后端推理 + 护栏 + 回退",保证任何异常/超时/护栏不过都安全
//     回退到传入的规则版文本 —— 绝不丢字、绝不卡住插入。
//   · 挂接点见 AppDelegate.corrected():Defiller 之后、ChineseITN 之前(isFinal 时)。
//   · 推理是 async:绝不能同步阻塞主线程/插入路径。
//
// 评估定稿(projects/Vibe_XASR/refiner_eval/):只做「去口癖 + 改口」,不做分段
// (0.6B 分段做不好);它是「润色」不是「忠实转写」,会改写口语,故默认关 + Beta 角标。
//
// 注:本文件并发隔离(Swift 6 strict concurrency)需本地 `swift build` 核对微调;
//     纯逻辑的 `Guardrails` 不涉并发,可独立编译测试。

/// 润色后端协议。实现负责加载模型 + 单次整段推理。
/// Sendable:推理在后台线程,实现需自行保证线程安全(actor,或 final class + 锁 / @unchecked Sendable)。
protocol RefinerBackend: AnyObject, Sendable {
    /// 模型是否已加载就绪。未就绪时门面直接回退原文(不等待)。
    var isReady: Bool { get }
    /// 对整段文本润色一次;返回 nil 表示后端无法处理(回退原文)。
    /// 入参 system 为定稿指令,text 为待润色文本。
    func refine(system: String, text: String) async -> String?
    /// 后端是否会在正文后追加「不确定词列表」(CPM5 本地模型的自报机制)。
    /// 门面据此决定是否剥离该列表;云端模型不会,用默认 false。
    var emitsUncertainList: Bool { get }
}

extension RefinerBackend {
    var emitsUncertainList: Bool { false }
}

/// AI 润色(Beta)门面。线程模型:门面在 MainActor,真正推理在后端的后台线程。
@MainActor
final class Refiner {
    static let shared = Refiner()

    /// 当前后端(默认 llama.cpp;懒加载)。nil = 未配置/模型未下载 → polish 为安全 no-op。
    var backend: RefinerBackend?

    /// 推理超时(秒)。超时即回退规则版。
    var timeout: TimeInterval = 4.0

    /// 指令构建器。本地=固定 systemPrompt;云端=AppDelegate 按配置(处理项/模板/热词,已替换
    /// {{hotwords}}/{{date}})拼成、可含 {{transcript}} 占位符,由后端在 refine 时替换为转写。
    var systemProvider: @MainActor () -> String = { Refiner.systemPrompt }

    /// CPM5(MiniCPM5-1B)官方固定 system prompt —— 开发者 corrector.py 原文,模型 SFT 即按此训练。
    /// 输出格式 `corrected_text<KEY>[词1、词2]`(<KEY> 后为不确定词,见 stripUncertainList)。
    /// ⚠️ 必须原样发送:不带它,模型走分布外退化路径(分隔符变 `<font>`/`<center>` 残渣、质量下降)。
    static let systemPrompt =
        "你是流式ASR后处理助手。输入ASR原始识别文本，输出修正后的规范文本。"
        + "去口癖、纠错字、加标点，规范书写格式，"
        + "保留全部语义不捏造。不确定的词在末尾标注<KEY>[词1、词2]。"
        + "直接输出结果，不要解释。"

    /// 入口:对最终文本润色。失败/超时/护栏不过 → 返回原文(安全回退)。
    /// - Parameter text: 已过规则链(Defiller 之后、ITN 之前)的整段文本。
    func polish(_ text: String) async -> String {
        guard let backend, backend.isReady, Refiner.shouldRun(text) else { return text }
        let sys = systemProvider()
        let raw = await Refiner.race(timeout: timeout) { await backend.refine(system: sys, text: text) }
        guard let raw, !raw.isEmpty else { return text }                 // 超时/空 → 回退
        var out = Refiner.stripWrapping(raw)
        if backend.emitsUncertainList { out = Refiner.stripUncertainList(out) }  // CPM5:剥掉尾部不确定词列表,绝不进插入路径
        guard !out.isEmpty, Guardrails.accept(src: text, out: out) else { return text }  // 护栏不过 → 回退
        return out
    }

    /// 触发判断:太短就别跑(省延迟、降风险)。够长才送后端。
    static func shouldRun(_ text: String) -> Bool {
        text.trimmingCharacters(in: .whitespacesAndNewlines).count >= 6
    }

    /// 清洗模型输出:剥掉偶发的包裹引号与残留 <think>…</think>。
    static func stripWrapping(_ s: String) -> String {
        var t = s.trimmingCharacters(in: .whitespacesAndNewlines)
        if let r = t.range(of: "<think>"), let e = t.range(of: "</think>"), r.lowerBound < e.upperBound {
            t.removeSubrange(r.lowerBound..<e.upperBound)
            t = t.trimmingCharacters(in: .whitespacesAndNewlines)
        }
        let quotes: Set<Character> = ["\"", "“", "”", "「", "」", "'"]
        if let f = t.first, quotes.contains(f) { t.removeFirst() }
        if let l = t.last, quotes.contains(l) { t.removeLast() }
        return t.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    /// 剥离模型在正文后追加的「不确定词列表」,只保留正文(列表本期丢弃,未来做兜底再用)。
    /// 官方格式(带固定 system prompt)为 `corrected_text<KEY>[词1、词2]` —— 按分隔符 `<KEY>` 切。
    /// 兜底:个别量化输出可能无 `<KEY>` 而残留尾部 `<…>` 块(分布外退化),再按尾部 `<…` 块剥一道;
    /// 语音正文几乎不含 `<`,误伤极低。
    static let uncertainKeySep = "<KEY>"
    static func stripUncertainList(_ s: String) -> String {
        if let r = s.range(of: uncertainKeySep) {
            return String(s[..<r.lowerBound]).trimmingCharacters(in: .whitespacesAndNewlines)
        }
        guard let r = s.range(of: #"\s*<[^<\n]{0,100}$"#, options: .regularExpression) else { return s }
        return String(s[..<r.lowerBound]).trimmingCharacters(in: .whitespacesAndNewlines)
    }

    /// 超时竞速:先到者胜;超时分支返回 nil → 调用方回退。
    /// （并发隔离细节需本地编译核对;后端 refine 应自行尊重取消以免后台泄漏。）
    nonisolated static func race(timeout: TimeInterval,
                                 _ op: @escaping @Sendable () async -> String?) async -> String? {
        await withTaskGroup(of: String?.self) { group in
            group.addTask { await op() }
            group.addTask {
                try? await Task.sleep(nanoseconds: UInt64(timeout * 1_000_000_000))
                return nil
            }
            let first = await group.next() ?? nil
            group.cancelAll()
            return first
        }
    }
}

/// 护栏:防止 refiner 丢信息/乱改。任一不通过 → 调用方回退原文。
/// 偏宽策略(CPM5):模型会主动 ITN(forty two→42)+ 改口,机械保「英文/数字全留」会误杀这些
/// 合法改写,故只留两道——① prompt-leak(复述指令)② 长度相对原文骤降(疑似丢信息)。
/// 风险词的复核交给模型自报的「不确定列表」(本期先剥离,后续可接强模型/联网兜底)。纯逻辑、无并发。
enum Guardrails {
    /// 输出比原文短超过该比例 → 判「删太多」回退。改口可合法删掉前半句(实测可达 ~60%),
    /// 故放宽到 0.7,只拦更极端的丢失,避免误伤改口。
    static let maxShrink: Double = 0.7

    static func accept(src: String, out: String) -> Bool {
        guard !looksLikePromptLeak(src: src, out: out) else { return false }  // 模型复述了指令 → 回退
        let shrink = 1.0 - Double(out.count) / Double(max(src.count, 1))
        return shrink <= maxShrink                          // 长度骤降兜底
    }

    /// prompt 泄漏检测:0.6B 小模型有时把 system 指令复述进输出(尤其短句、无可整理处)。
    /// 输出含指令特征词、而原文没有 → 判为泄漏,回退原文。
    static func looksLikePromptLeak(src: String, out: String) -> Bool {
        let markers = ["流式ASR", "后处理助手", "ASR原始", "纠错字", "不捏造",
                       "标注<KEY>", "直接输出结果", "不要解释"]
        return markers.contains { out.contains($0) && !src.contains($0) }
    }

}

File: projects/Vibe_XASR/windows_build/src/VibeXASR.Windows/Refine/Refiner.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace VibeXASR.Windows.Refine;

/// <summary>A refinement backend (cloud / local). Text in → polished out. Port of RefinerBackend.</summary>
internal interface IRefinerBackend
{
    /// <summary>Whether the backend is configured + ready. Not-ready → facade returns the input untouched.</summary>
    bool IsReady { get; }
    /// <summary>Refine the whole text once. Returns null when the backend can't handle it (caller falls back).</summary>
    Task<string?> RefineAsync(string system, string text, CancellationToken ct);
    /// <summary>Whether the backend appends an "uncertain words" list after the body — the CPM5 local model's
    /// self-report (`…&lt;KEY&gt;[词1、词2]`). The facade strips it before insert and relaxes the guardrails
    /// (CPM5 does its own ITN/改口, so the English/digit-kept checks would wrongly reject it). Cloud → false.</summary>
    bool EmitsUncertainList => false;
}

/// <summary>
/// AI refinement (Beta) — backend-agnostic facade. Faithful port of macOS Refiner.swift.
/// Coordinates backend inference + guardrails + fallback so ANY error/timeout/guardrail-reject safely
/// falls back to the rule-version text — never drops characters, never blocks insertion.
/// Hook point: TrayApp.OnFinal, after the rule chain (defiller/ITN/snippets), before insert.
/// </summary>
internal static class Refiner
{
    /// <summary>Current backend (null = unconfigured → PolishAsync is a safe no-op).</summary>
    public static IRefinerBackend? Backend;

    /// <summary>Inference timeout (seconds). Timeout → fall back to the rule version. Cloud is slow (3–8s typical).</summary>
    public static double TimeoutSeconds = 25.0;

    /// <summary>Instruction builder. Cloud = TrayApp builds it from the config (mods/template/hotwords, with
    /// {{hotwords}}/{{date}} already filled, may contain {{transcript}} for the backend to fill).</summary>
    public static Func<string> SystemProvider = () => SystemPrompt;

    /// <summary>Default fixed instruction: remove fillers + keep the final restatement; never touch numbers/English/names.</summary>
    public const string SystemPrompt =
        "你是语音转写(ASR)文本的整理助手。只做两件事:" +
        "① 删除口癖词(嗯、呃、那个、就是、然后那个 等)与明显重复;" +
        "② 若说话人中途改口(如「周二…不对周三」),只保留最终说法。" +
        "然后补全标点。若原文没有需要修改的,就原样输出原文,不要补充任何内容。" +
        "严禁复述本说明或任何指令,严禁解释,严禁翻译,严禁改动数字、英文与专有名词。" +
        "只输出整理后的文本本身。";

    /// <summary>CPM5(MiniCPM5-1B)官方固定 system prompt —— 开发者 corrector.py 原文,模型 SFT 即按此训练。
    /// 输出格式 <c>corrected_text&lt;KEY&gt;[词1、词2]</c>(&lt;KEY&gt; 后为不确定词,见 <see cref="StripUncertainList"/>)。
    /// ⚠️ 必须原样发送:不带它,模型走分布外退化路径(分隔符变 <c>&lt;font&gt;/&lt;center&gt;</c> 残渣、质量下降)。
    /// Used only by the LOCAL backend (set as SystemProvider when LocalRefiner is active).</summary>
    public const string CpmSystemPrompt =
        "你是流式ASR后处理助手。输入ASR原始识别文本，输出修正后的规范文本。" +
        "去口癖、纠错字、加标点，规范书写格式，" +
        "保留全部语义不捏造。不确定的词在末尾标注<KEY>[词1、词2]。" +
        "直接输出结果，不要解释。";

    public static bool Active => Backend?.IsReady == true;

    /// <summary>Polish final text. Failure/timeout/guardrail-reject → returns the input (safe fallback).</summary>
    public static async Task<string> PolishAsync(string text)
    {
        var backend = Backend;
        if (backend is null || !backend.IsReady || !ShouldRun(text)) return text;
        var sys = SystemProvider();
        string? raw;
        try
        {
            using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(TimeoutSeconds));
            raw = await backend.RefineAsync(sys, text, cts.Token).ConfigureAwait(false);
        }
        catch { raw = null; }                                   // timeout/cancel/error → fall back
        if (string.IsNullOrEmpty(raw)) return text;
        var outp = StripWrapping(raw!);
        if (backend.EmitsUncertainList) outp = StripUncertainList(outp);   // CPM5: drop the tail list, never insert it
        if (string.IsNullOrEmpty(outp) || !Guardrails.Accept(text, outp, backend.EmitsUncertainList)) return text;  // guardrail reject → fall back
        return outp;
    }

    /// <summary>Trigger gate: too short → skip (saves latency + risk). Port of shouldRun (≥6 chars).</summary>
    public static bool ShouldRun(string text) => text.Trim().Length >= 6;

    /// <summary>Clean model output: strip stray wrapping quotes and leftover &lt;think&gt;…&lt;/think&gt;.</summary>
    public static string StripWrapping(string s)
    {
        var t = s.Trim();
        int open = t.IndexOf("<think>", StringComparison.OrdinalIgnoreCase);
        int close = t.IndexOf("</think>", StringComparison.OrdinalIgnoreCase);
        if (open >= 0 && close >= 0 && close > open)
            t = (t.Substring(0, open) + t.Substring(close + "</think>".Length)).Trim();
        var quotes = new HashSet<char> { '"', '“', '”', '「', '」', '\'' };
        if (t.Length > 0 && quotes.Contains(t[0])) t = t.Substring(1);
        if (t.Length > 0 && quotes.Contains(t[^1])) t = t.Substring(0, t.Length - 1);
        return t.Trim();
    }

    /// <summary>Strip the "uncertain words" list the CPM5 model appends after the body, keeping only the body.
    /// Official format (with the fixed CPM5 system prompt) is <c>corrected_text&lt;KEY&gt;[词1、词2]</c> — split on
    /// <c>&lt;KEY&gt;</c>. Fallback: a quantized output may drop <c>&lt;KEY&gt;</c> and leave a trailing <c>&lt;…&gt;</c>
    /// block (out-of-distribution); strip a short trailing <c>&lt;…</c> block too (speech bodies rarely contain '&lt;',
    /// so the false-positive risk is tiny). Port of macOS Refiner.stripUncertainList.</summary>
    public const string UncertainKeySep = "<KEY>";
    public static string StripUncertainList(string s)
    {
        int i = s.IndexOf(UncertainKeySep, StringComparison.Ordinal);
        if (i >= 0) return s.Substring(0, i).Trim();
        var m = System.Text.RegularExpressions.Regex.Match(s, @"\s*<[^<\n]{0,100}$");
        return m.Success ? s.Substring(0, m.Index).Trim() : s;
    }
}

/// <summary>Guardrails: prevent the refiner from losing info / mangling. Any reject → caller falls back.
/// Faithful port of macOS Guardrails. Pure logic.</summary>
internal static class Guardrails
{
    /// <summary>Output shorter than the source by more than this ratio → "deleted too much" reject. A genuine
    /// restatement can legitimately drop the first half (~60% measured), so 0.7 only catches extreme loss.</summary>
    public const double MaxShrink = 0.7;

    /// <param name="loose">CPM5/local: the model actively does ITN (forty two→42, 一百二十三→123) + 改口, so the
    /// mechanical "keep every English/digit run" checks would wrongly reject these legitimate rewrites. macOS
    /// drops them for CPM5 and keeps only ① prompt-leak ② length-collapse. Cloud (strict) keeps all four.</param>
    public static bool Accept(string src, string outp, bool loose = false)
    {
        if (LooksLikePromptLeak(src, outp)) return false;       // model parroted the instruction → reject
        if (!loose)
        {
            if (!EnglishKept(src, outp)) return false;          // every English word must survive
            if (!DigitsKept(src, outp)) return false;           // every Arabic-digit run must survive
        }
        var shrink = 1.0 - (double)outp.Length / Math.Max(src.Length, 1);
        return shrink <= MaxShrink;                             // length-collapse backstop
    }

    /// <summary>Prompt-leak detection: small models sometimes echo the system instruction into the output.
    /// Output contains an instruction marker that the source did not → treat as leak, fall back.
    /// Covers both the cloud rule-prompt markers and the CPM5 local-prompt markers (harmless to check both).</summary>
    public static bool LooksLikePromptLeak(string src, string outp)
    {
        string[] markers = { "语音转写", "整理助手", "口癖", "改口", "保留最终说法",
                             "明显重复", "原样输出", "本说明", "ASR文本", "ASR)",
                             // CPM5 local-prompt markers (port of macOS Guardrails)
                             "流式ASR", "后处理助手", "ASR原始", "纠错字", "不捏造",
                             "标注<KEY>", "直接输出结果", "不要解释" };
        return markers.Any(m => outp.Contains(m) && !src.Contains(m));
    }

    /// <summary>Every ASCII English word in the source must appear in the output (case-insensitive).</summary>
    public static bool EnglishKept(string src, string outp) =>
        AsciiRuns(src.ToLowerInvariant(), char.IsLetter).IsSubsetOf(AsciiRuns(outp.ToLowerInvariant(), char.IsLetter));

    /// <summary>Every Arabic-digit run in the source must appear in the output.</summary>
    public static bool DigitsKept(string src, string outp) =>
        AsciiRuns(src, char.IsDigit).IsSubsetOf(AsciiRuns(outp, char.IsDigit));

    /// <summary>Contiguous runs of ASCII letters / digits.</summary>
    private static HashSet<string> AsciiRuns(string s, Func<char, bool> keep)
    {
        var set = new HashSet<string>();
        var cur = new StringBuilder();
        foreach (var ch in s)
        {
            if (ch < 128 && keep(ch)) cur.Append(ch);
            else if (cur.Length > 0) { set.Add(cur.ToString()); cur.Clear(); }
        }
        if (cur.Length > 0) set.Add(cur.ToString());
        return set;
    }
}
